File: apps/app.music/server/src/main/scala/com/nublic/app/music/server/filewatcher/Images.scala
package com.nublic.app.music.server.filewatcher

import com.nublic.app.music.server.model.Artist
import com.nublic.app.music.server.model.Album
import com.echonest.api.{v4 => E}
import com.echonest.api.v4.util.Commander
import java.net.URL
import org.apache.commons.io.FileUtils
import com.nublic.app.music.server.MusicFolder
import java.io.File
import java.io.FileReader
import org.im4java.core.ConvertCmd
import org.im4java.core.IMOperation
import net.liftweb.json._
import org.apache.commons.io.IOUtils
import java.net.URI
import org.jaudiotagger.audio.AudioFileIO
import java.io.FileOutputStream

object Images {
  
  val ECHONEST_API_KEY = "UR4VKX7JXDXAULIWB"
  implicit val formats = Serialization.formats(NoTypeHints)
  
  def ensure(id: Long, folderer: Long => File, getter: File => Unit): Unit = {
    val folder = folderer(id)
	// Create folder if it does not exist
    if (!folder.exists()) {
      folder.mkdirs()
    }
    // Download original image from web
    val original = new File(folder, MusicFolder.ORIGINAL_FILENAME)
    if (!original.exists()) {
      getter(original)
    }
    // Create thumbnail
    val thumb = new File(folder, MusicFolder.THUMBNAIL_FILENAME)
    if (original.exists() && !thumb.exists()) {
      val magick = new ConvertCmd()
      val op = new IMOperation() 
      op.addImage(original.getAbsolutePath())
      op.resize(MusicFolder.ARTIST_THUMBNAIL_SIZE, MusicFolder.ARTIST_THUMBNAIL_SIZE)
      op.addImage(thumb.getAbsolutePath())
      magick.run(op)
    }
  }
    
  def ensureArtist(a: Artist) = 
    ensure(a.id, l => MusicFolder.getArtistFolder(l), getArtistImage(a))
  
  def getArtistImage(a: Artist)(f: File) = {
    try {
      val api = new E.EchoNestAPI(ECHONEST_API_KEY)
      val artist = api.newArtistByName(a.name)
      val image_infos = artist.getImages()
      if (!image_infos.isEmpty()) {
        val image_url = new URL(image_infos.get(0).getURL())
        FileUtils.copyURLToFile(image_url, f)
      }
    } catch {
      case _: Throwable => { /* Do nothing */ }
    }
  }
  
  def ensureAlbum(mp3_file: File, ab: Album, ar: Option[Artist]) = 
    ensure(ab.id, l => MusicFolder.getAlbumFolder(l), getAlbumImage(mp3_file, ab, ar))
  
  def getAlbumImage(mp3_file: File, ab: Album, ar: Option[Artist])(f: File) {
    for (getter <- List(getAlbumImageFilesystem(mp3_file)_, getAlbumImageEmbedded(mp3_file)_, getAlbumImageDiscogs(ab, ar)_)) {
      if (!f.exists()) {
        getter(f)
      }
    }
  }
    
  def getAlbumImageFilesystem(mp3_f: File)(image_f: File) = {
    val regex = """(cover|folder|front|album|albumart)\.(jpg|jpeg|png|bmp)"""
    mp3_f.getParentFile().listFiles().toList.filterNot(_.isDirectory).filter(f => f.getName().matches(regex)) match {
      case cover_file :: _ => FileUtils.copyFile(cover_file, image_f)
      case _               => { /* Do nothing */ }
    }
  }
  
  def getAlbumImageEmbedded(mp3_f: File)(image_f: File) = {
    try {
      val audio = AudioFileIO.read(mp3_f)
      val tag = audio.getTag()
      val artwork = tag.getFirstArtwork()
      if (artwork != null) {
        val stream = new FileOutputStream(image_f)
        stream.write(artwork.getBinaryData())
        stream.flush()
        stream.close()
      }
    } catch {
      case _: Throwable => { /* Do nothing */ }
    }
  }
  
  def getAlbumImageDiscogs(ab: Album, ar: Option[Artist])(f: File) = {
    try {
      val search = "f=json&type=releases&q=" + ab.name + " " + ar.map(_.name).getOrElse("")
      val json_uri = new URI("http", "api.discogs.com", "/search", search, null)
      val json_url = json_uri.toURL()
      val con = json_url.openConnection()
      con.addRequestProperty("Accept-Encoding", "gzip")
      val in = con.getInputStream()
      var encoding = con.getContentEncoding()
      encoding = if (encoding == null) "UTF-8" else encoding
      val body = IOUtils.toString(in, encoding)
      
      val response = parse(body).extract[DiscogResponse]
      if (!response.resp.search.searchresults.results.isEmpty) {
        val result = response.resp.search.searchresults.results.head
        val image_url = new URL(result.thumb)
        FileUtils.copyURLToFile(image_url, f)
      }
    } catch {
      case _: Throwable => { /* Do nothing */ }
    }
  }
  
  case class DiscogResult(thumb: String, title: String)
  case class DiscogResults(numResults: String, start: String, end: String, results: List[DiscogResult])
  case class DiscogSearch(searchresults: DiscogResults)
  case class DiscogResp(status: Boolean, version: String, search: DiscogSearch)
  case class DiscogResponse(resp: DiscogResp)
}

File: apps/app.music/server-python/src/nublic_app_music_server/images.py
import os
import os.path
from pgmagick.api import Image
import re
import requests
import shutil
import json

from pyechonest import config
import pyechonest.artist

# Get the cache path from environment
try:
    ROOT_CACHE_MUSIC_FOLDER = os.environ["MUSIC_CACHE_FOLDER"]
except KeyError:
    ROOT_CACHE_MUSIC_FOLDER = "/var/nublic/cache/music"

ARTISTS_FOLDER = os.path.join(ROOT_CACHE_MUSIC_FOLDER, "artists")
ALBUMS_FOLDER = os.path.join(ROOT_CACHE_MUSIC_FOLDER, "albums")
ORIGINAL_FILENAME = "orig"
THUMBNAIL_FILENAME = "thumbnail.png"
THUMBNAIL_SIZE = 96

import logging
log = logging.getLogger(__name__)


def get_artist_folder(artist_id):
    return os.path.join(ARTISTS_FOLDER, unicode(artist_id).encode('utf8'))


def get_album_folder(artist_id):
    return os.path.join(ALBUMS_FOLDER, unicode(artist_id).encode('utf8'))


def _ensure(id_, folderer, getter):
    folder = folderer(id_)
    # Create folder if it does not exists
    if not os.path.exists(folder):
        os.makedirs(folder)
    # Download image
    original = os.path.join(folder, ORIGINAL_FILENAME)
    if not os.path.exists(original):
        getter(original)
    # Create thumbnail
    thumb = os.path.join(folder, THUMBNAIL_FILENAME)
    if os.path.exists(original) and not os.path.exists(thumb):
        img = Image(original)
        img.scale((THUMBNAIL_SIZE, THUMBNAIL_SIZE))
        img.write(thumb)


def request_several_times(url, times=3, timeout=5):
    for i in range(0, times):
        try:
            r = requests.get(url, timeout=timeout)
            return r
        except requests.Timeout:
            log.info("Timeout achieved calling %s", url)
    log.info("Requesting %d times failed", times)
    return None


def ensure_artist_image(id_, name):
    _ensure(id_, get_artist_folder, lambda path: get_artist_image(id_,
            name, path))


def get_artist_image(id_, name, place):
    # Get from Echo Nest
    try:
        config.ECHO_NEST_API_KEY = 'UR4VKX7JXDXAULIWB'
        results = pyechonest.artist.search(name=name)
        if results:
            artist = results[0]
            images = artist.get_images(results=1)
            if images:
                image_url = images[0][u'url']
                r = request_several_times(image_url)
                if r is not None and r.status_code == 200:
                    f = open(place, 'w')
                    f.write(r.content)
                    f.close()
                    return True
    except:
        log.exception("Exception detected while getting an artist image")
    return False


def ensure_album_image(id_, file_, album_name, artist_name):
    _ensure(id_, get_album_folder, lambda path: get_album_image(id_,
            file_, album_name, artist_name, path))


def get_album_image(id_, file_, album_name, artist_name, place):
    """
    file_ must be a byte string in utf8 encoding
    """
    # Try to get from filesystem
    parent_f, _ = os.path.split(file_)
    for f in os.listdir(parent_f):
        r = re.match(
            r"(cover|folder|front|album|albumart)\.(jpg|jpeg|png|bmp)", f)
        if r is not None:
            shutil.copyfile(os.path.join(parent_f, f), place)
            return True
    # Try to get from discogs
    try:
        if artist_name is None:
            search = album_name
        else:
            search = album_name + u' ' + artist_name
        discogs_params = {'f': 'json',
                          'type': 'releases', 'q': search}
        r = requests.get(
            'http://api.discogs.com/search', params=discogs_params)
        log.debug("Gets request %s", r)
        response = json.loads(r.content)
        if response is not None and response['resp']['status']:
            img_results = response['resp']['search']['searchresults']['results']
            log.debug("img_results %s", str(img_results))
            if img_results:
                image_url = img_results[0]['thumb']
                log.debug("img_url %s", str(image_url))
                r = request_several_times(image_url)
                if r is not None:
                    f = open(place, 'w')
                    f.write(r.content)
                    f.close()
                    return True
    except BaseException:
        log.exception("Exception detected getting album from api.discogs.com")
    return False
